cli_tools/common/utils/logging/firelog_entry.go

```diff
@@ -0,0 +1,207 @@
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package logging
+
+import (
+	"encoding/json"
+	"fmt"
+	"runtime"
+	"time"
+)
+
+// LogRequest is a server-side pre-defined data structure
+type LogRequest struct {
+	ClientInfo    ClientInfo `json:"client_info"`
+	LogSourceName string     `json:"log_source_name"`
+	RequestTimeMs int64      `json:"request_time_ms"`
+	LogEvent      []LogEvent `json:"log_event"`
+}
+
+// ClientInfo is a server-side pre-defined data structure
+type ClientInfo struct {
+	ClientType        string            `json:"client_type"`
+	DesktopClientInfo map[string]string `json:"desktop_client_info"`
+}
+
+// LogEvent is a server-side pre-defined data structure
+type LogEvent struct {
+	EventTimeMs         int64  `json:"event_time_ms"`
+	SourceExtensionJSON string `json:"source_extension_json"`
+}
+
+// LogResponse is a server-side pre-defined data structure
+type LogResponse struct {
+	NextRequestWaitMillis int64                `json:"nextRequestWaitMillis,string"`
+	LogResponseDetails    []LogResponseDetails `json:"logResponseDetails"`
+}
+
+// LogResponseDetails is a server-side pre-defined data structure
+type LogResponseDetails struct {
+	ResponseAction ResponseAction `json:"responseAction"`
+}
+
+// ResponseAction is a server-side pre-defined data structure
+type ResponseAction string
+
+const (
+	// ResponseActionUnknown - If the client sees this, it should delete the LogRequest (not retry).
+	// It may indicate that a new response action was added, which the client
+	// doesn't yet understand.  (Deleting rather than retrying will prevent
+	// infinite loops.)  The server will do whatever it can to prevent this
+	// occurring (by not indicating an action to clients that are behind the
+	// requisite version for the action).
+	ResponseActionUnknown ResponseAction = "RESPONSE_ACTION_UNKNOWN"
+	// RetryRequestLater - The client should retry the request later, via normal scheduling.
+	RetryRequestLater ResponseAction = "RETRY_REQUEST_LATER"
+	// DeleteRequest - The client should delete the request.  This action will apply for
+	// successful requests, and non-retryable requests.
+	DeleteRequest ResponseAction = "DELETE_REQUEST"
+)
+
+func constructLogRequest(event *ComputeImageToolsLogExtension) ([]byte, error) {
+	if event == nil {
+		return nil, fmt.Errorf("won't log a nil event")
+	}
+	eventStr, err := json.Marshal(event)
+	if err != nil {
+		return nil, err
+	}
+
+	now := time.Now().Unix()
+	req := LogRequest{
+		ClientInfo: ClientInfo{
+			ClientType:        "DESKTOP",
```

Comment:
Consider adding your own client type rather than labeling yourself as a canonical client.

Reply:
OK I plan to use "COMPUTE_IMAGE_TOOLS" as the new client name